src/modules/flow/string/string-ascii.c

```diff
@@ -236,6 +238,132 @@ string_compare(struct sol_flow_node *node,
         SOL_FLOW_NODE_TYPE_STRING_COMPARE__OUT__OUT, result);
 }
 
+struct string_slice_data {
+    struct sol_flow_node *node;
+    char *str;
+    int idx[2];
+};
+
+static int
+get_slice_idx_by_port(const struct sol_flow_packet *packet,
+    uint16_t port,
+    struct string_slice_data *mdata)
+{
+    int32_t in_value;
+    int r;
+
+    r = sol_flow_packet_get_irange_value(packet, &in_value);
+    SOL_INT_CHECK(r, < 0, r);
+
+    mdata->idx[port] = in_value;
+
+    return 0;
+}
+
+static int
+slice_do(struct string_slice_data *mdata)
+{
+    struct sol_str_slice slice;
+    int32_t start = mdata->idx[0], end = mdata->idx[1],
+        len = strlen(mdata->str);
+
+#define BOUND_CHECK(_val) \
+    do { \
+        if (_val >= 0) { \
+            if (!len && _val) \
+                _val = 0; \
+            else if (_val > len) \
+                _val = len; \
+        } else if (_val < -len) { \
+            _val = 0; \
+        } else { \
+            _val = len + _val; \
+        } \
+    } while (0)
```

Comment:
I'd say don't create a macro, but a `static inline` function in `sol-util.h` to clamp a number within 2 values.

Then you should compute the value and clamp it:
```c
static inline int32_t
sol_util_int32_clamp(int32_t start, int32_t end, int32_t value)
{
    if (value < start)
        return start;
    if (value > end)
        return end;
    return value;
}

static int
slice_do(...) {
    ...
    if (start < 0) start = len + start;
    if (end < 0) end = len + end;
    start = sol_util_int32_clamp(0, len, start);
    end = sol_util_int32_clamp(0, len, end);
```